FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1494 @@
+                                   uint16_t topicFilterLength );
+
+/**
+ * @brief Populate the parameters of a #Command_t
+ *
+ * @param[in] xCommandType Type of command.
+ * @param[in] pxContext Context and necessary structs for command.
+ * @param[in] xCallback Callback for when command completes.
+ * @param[out] pxCommand Pointer to initialized command.
+ *
+ * @return `true` if all necessary structs for the command exist in pxContext,
+ * else `false`
+ */
+static bool prvCreateCommand( CommandType_t xCommandType,
+                              CommandContext_t * pxContext,
+                              CommandCallback_t xCallback,
+                              Command_t * pxCommand );
+
+/**
+ * @brief Add a command to the global command queue.
+ *
+ * @param[in] pxCommand Pointer to command to copy to queue.
+ */
+static void prvAddCommandToQueue( Command_t * pxCommand );
+
+/**
+ * @brief Copy an incoming publish to a response queue.
+ *
+ * @param[in] pPublishInfo Info of incoming publish.
+ * @param[in] pResponseQueue Queue to which the publish is copied.
+ */
+static void prvCopyPublishToQueue( MQTTPublishInfo_t * pPublishInfo,
+                                   QueueHandle_t pResponseQueue );
+
+/**
+ * @brief Process a #Command_t.
+ *
+ * @note This demo does not check existing subscriptions before sending a
+ * SUBSCRIBE or UNSUBSCRIBE packet. If a subscription already exists, then
+ * a SUBSCRIBE packet will be sent anyway, and if multiple tasks are subscribed
+ * to a topic filter, then they will all be unsubscribed after an UNSUBSCRIBE.
+ *
+ * @param[in] pxCommand Pointer to command to process.
+ *
+ * @return status of MQTT library API call.
+ */
+static MQTTStatus_t prvProcessCommand( Command_t * pxCommand );
+
+/**
+ * @brief Dispatch incoming publishes and acks to response queues and
+ * command callbacks.
+ *
+ * @param[in] pMqttContext MQTT Context
+ * @param[in] pPacketInfo Pointer to incoming packet.
+ * @param[in] pDeserializedInfo Pointer to deserialized information from
+ * the incoming packet.
+ */
+static void prvSubscriptionManager( MQTTContext_t * pMqttContext,
+                                    MQTTPacketInfo_t * pPacketInfo,
+                                    MQTTDeserializedInfo_t * pDeserializedInfo );
+
+/**
+ * @brief Process commands from the command queue in a loop.
+ *
+ * This demo requires a process loop command to be enqueued before calling this
+ * function, and will re-add a process loop command every time one is processed.
+ * This demo will exit the loop after receiving an unsubscribe operation.
+ */
+static void prvCommandLoop();
+
+/**
+ * @brief Common callback for commands in this demo.
+ *
+ * This callback marks the command as complete and notifies the calling task.
+ *
+ * @param[in] pxContext Context of the initial command.
+ */
+static void prvCommandCallback( CommandContext_t * pxContext );
+
+/**
+ * @brief The task used to create various publish operations.
```

Comment:
I think a bit more detail about what the task is actually doing will make the code more readable. The same description can be copied to demo description at the start of this file as well. Applies to all 3 tasks in this demo.

Reply:
Added more comments

Reply:
Looks good. Thanks